### PR TITLE
Animation fixes

### PR DESCRIPTION
**Pull request by ludvigarlebrink**

### CODE REVIEW

File: src/Editor/Util/AssetConverterSkeleton.cpp [hunk 1]
```diff
@@ -116,19 +148,58 @@ const std::string& AssetConverterSkeleton::GetErrorString() {
...
+            if (child->mName.C_Str() == bones[j]) {
+                children.push_back(child->mName.C_Str());
+                ids.push_back(0);
+                ++currentId;
+                parents.push_back(-1);
+                BoneRecursive(child, 0);
+            }
+        }
+
+        if (!foundBone)
```
Comment (NinjaDanz3r) [inline src/Editor/Util/AssetConverterSkeleton.cpp]: MINOR Condition '!foundBone' is always true [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AknownConditionTrueFalse)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 9 issues
* ![MINOR][MINOR] 9 minor

Watch the comments in this conversation to review them.

#### 8 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [RenderManager.cpp#L508](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/3c3bb70d6df0bcce39498809e1f905a3e076eede/src/Engine/Manager/RenderManager.cpp#L508): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L526](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/3c3bb70d6df0bcce39498809e1f905a3e076eede/src/Engine/Manager/RenderManager.cpp#L526): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L545](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/3c3bb70d6df0bcce39498809e1f905a3e076eede/src/Engine/Manager/RenderManager.cpp#L545): The function 'GetLenses' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L565](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/3c3bb70d6df0bcce39498809e1f905a3e076eede/src/Engine/Manager/RenderManager.cpp#L565): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L583](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/3c3bb70d6df0bcce39498809e1f905a3e076eede/src/Engine/Manager/RenderManager.cpp#L583): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L602](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/3c3bb70d6df0bcce39498809e1f905a3e076eede/src/Engine/Manager/RenderManager.cpp#L602): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L624](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/3c3bb70d6df0bcce39498809e1f905a3e076eede/src/Engine/Manager/RenderManager.cpp#L624): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/Util/AssetConverterSkeleton.cpp [hunk 2]
```diff
@@ -37,34 +36,50 @@ bool AssetConverterSkeleton::Convert(const char* filepath, const char* destinati
...
         Animation::Skeleton skeleton;
         for (unsigned int i = 0; i < aScene->mAnimations[0]->mNumChannels; ++i) {
-            aiNodeAnim* channel = aScene->mAnimations[0]->mChannels[i];
+            auto boneIndex = 0;
+            for (auto j = 0; j < aScene->mAnimations[0]->mNumChannels; ++j) {
+                if (aScene->mAnimations[0]->mChannels[j]->mNodeName.C_Str() == bones[i]) {
+                    boneIndex = j;
+                    break;
+                }
+            }
```
Comment (Sajdkick) [inline src/Editor/Util/AssetConverterSkeleton.cpp]: Unnecessary {}

File: src/Editor/Util/AssetConverterSkeleton.cpp [hunk 3]
```diff
@@ -116,19 +148,58 @@ const std::string& AssetConverterSkeleton::GetErrorString() {
...
+        bool foundBone = false;
+        for (auto j = 0; j < bones.size(); ++j) {
+            if (child->mName.C_Str() == bones[j]) {
+                children.push_back(child->mName.C_Str());
+                ids.push_back(0);
+                ++currentId;
+                parents.push_back(-1);
+                BoneRecursive(child, 0);
+            }
+        }
```
Comment (Sajdkick) [inline src/Editor/Util/AssetConverterSkeleton.cpp]: Unnecessary {}

File: src/Editor/Util/AssetConverterSkeleton.cpp [hunk 4]
```diff
@@ -116,19 +148,58 @@ const std::string& AssetConverterSkeleton::GetErrorString() {
...
                 children.push_back(child->mName.C_Str());
-                parents.push_back(parent);
+                ids.push_back(thisId);
+
+                for (auto k = 0; k < bones.size(); ++k) {
+                    if (bones[k] == child->mParent->mName.C_Str()) {
+                        parents.push_back(k);
+                        break;
+                    }
+                }
```
Comment (Sajdkick) [inline src/Editor/Util/AssetConverterSkeleton.cpp]: Unnecessary {}

File: src/Editor/Util/AssetConverterSkeleton.cpp [hunk 5]
```diff
@@ -116,19 +148,58 @@ const std::string& AssetConverterSkeleton::GetErrorString() {
...
+
                 ++currentId;
-                ids.push_back(currentId);
-                break;
+                foundBone = true;
+                BoneRecursive(child, thisId);
             }
         }
```
Comment (Sajdkick) [inline src/Editor/Util/AssetConverterSkeleton.cpp]: Unnecessary {}

File: src/Engine/Animation/Skeleton.cpp [hunk 6]
```diff
@@ -74,6 +74,9 @@ void Skeleton::InitSkeleton() {
 
     // Setup rest of skeleton.
     for (unsigned int i = 1; i < skeletonBones.size(); ++i) {
+        if (skeletonBones[i]->parentId > 100)
```
Comment (Sajdkick) [inline src/Engine/Animation/Skeleton.cpp]: Add a comment to explain the 100?

File: src/Engine/Component/AnimationController.cpp [hunk 7]
```diff
@@ -88,8 +88,8 @@ void AnimationController::Animate(float deltaTime, Animation::AnimationControlle
     Animation::AnimationClip::Animation* anim = action->animationClip->animation;
     std::size_t size = skeleton->skeletonBones.size() > anim->numBones ? anim->numBones : skeleton->skeletonBones.size();
 
-    anim->currentFrame += deltaTime * 1.0f;
-    if (anim->currentFrame > 15.0f) {
+    anim->currentFrame += deltaTime * 30.0f;
+    if (anim->currentFrame > 119.0f) {
```
Comment (Sajdkick) [inline src/Engine/Component/AnimationController.cpp]: Add comment to explain the 119.0f?

File: src/Engine/Component/AnimationController.cpp [hunk 8]
```diff
@@ -102,12 +102,12 @@ void AnimationController::Animate(float deltaTime, Animation::AnimationControlle
     for (std::size_t i = 1; i < size; ++i) {
         Animation::AnimationClip::Bone* bone = &anim->bones[i];
 
-        if ((float)bone->rotationKeys[bone->currentKeyIndex + 1] > anim->currentFrame)
+        // Loop if the animation is very fast.
+        while ((float)bone->rotationKeys[bone->currentKeyIndex + 1] < anim->currentFrame) {
             ++bone->currentKeyIndex;
+        }
```
Comment (Sajdkick) [inline src/Engine/Component/AnimationController.cpp]: Unnecessary {}

File: src/Engine/Manager/RenderManager.cpp [hunk 9]
```diff
@@ -311,7 +311,7 @@ void RenderManager::RenderWorldEntities(World& world, const glm::mat4& viewMatri
             if (entity->IsKilled() || !entity->enabled)
                 continue;
 
-            if (mesh->geometry != nullptr && mesh->geometry->GetType() == Video::Geometry::Geometry3D::STATIC) {
+            if (mesh->geometry && mesh->geometry->GetType() == Video::Geometry::Geometry3D::STATIC) {
                 if (entity->GetComponent<Material>() != nullptr)
                     renderer->DepthRenderStaticMesh(mesh->geometry, viewMatrix, projectionMatrix, entity->GetModelMatrix());
             }
```
Comment (Sajdkick) [inline src/Engine/Manager/RenderManager.cpp]: Unnecessary {}

File: src/Engine/Manager/RenderManager.cpp [hunk 10]
```diff
@@ -325,7 +325,7 @@ void RenderManager::RenderWorldEntities(World& world, const glm::mat4& viewMatri
                 continue;
 
             Mesh* mesh = entity->GetComponent<Mesh>();
-            if (mesh->geometry != nullptr && mesh->geometry->GetType() == Video::Geometry::Geometry3D::SKIN) {
+            if (mesh && mesh->geometry && mesh->geometry->GetType() == Video::Geometry::Geometry3D::SKIN) {
                 if (entity->GetComponent<Material>() != nullptr)
                     renderer->DepthRenderSkinMesh(mesh->geometry, viewMatrix, projectionMatrix, entity->GetModelMatrix(), controller->bones);
             }
```
Comment (Sajdkick) [inline src/Engine/Manager/RenderManager.cpp]: Unnecessary {}

File: src/Engine/Component/AnimationController.cpp (hunk 7 above)
Comment (ludvigarlebrink) [reply src/Engine/Component/AnimationController.cpp]: This is hardcoded, I'm changing this to a non hardcoded value. And feel this code lacks comments overall so will be adding some more comments.

**Comment by ludvigarlebrink** [comment]
@Sajdkick Fixed requested changes.

**Comment by NinjaDanz3r** [comment]
iamheman

File: src/Editor/Util/AssetConverterSkeleton.cpp [hunk 11]
```diff
@@ -116,19 +150,55 @@ const std::string& AssetConverterSkeleton::GetErrorString() {
...
+            if (child->mName.C_Str() == bones[j]) {
+                children.push_back(child->mName.C_Str());
+                ids.push_back(0);
+                ++currentId;
+                parents.push_back(-1);
+                BoneRecursive(child, 0);
+            }
 
-        for (unsigned int j = 0; j < bones.size(); ++j) {
+        if (!foundBone)
```
Comment (NinjaDanz3r) [inline src/Editor/Util/AssetConverterSkeleton.cpp]: MINOR Condition '!foundBone' is always true [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AknownConditionTrueFalse)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 9 issues
* ![MINOR][MINOR] 9 minor

Watch the comments in this conversation to review them.

#### 8 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [RenderManager.cpp#L506](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6cf2e48c750d140b71857ae073e2aa19434d892b/src/Engine/Manager/RenderManager.cpp#L506): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L524](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6cf2e48c750d140b71857ae073e2aa19434d892b/src/Engine/Manager/RenderManager.cpp#L524): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L543](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6cf2e48c750d140b71857ae073e2aa19434d892b/src/Engine/Manager/RenderManager.cpp#L543): The function 'GetLenses' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L563](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6cf2e48c750d140b71857ae073e2aa19434d892b/src/Engine/Manager/RenderManager.cpp#L563): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L581](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6cf2e48c750d140b71857ae073e2aa19434d892b/src/Engine/Manager/RenderManager.cpp#L581): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L600](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6cf2e48c750d140b71857ae073e2aa19434d892b/src/Engine/Manager/RenderManager.cpp#L600): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L622](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/6cf2e48c750d140b71857ae073e2aa19434d892b/src/Engine/Manager/RenderManager.cpp#L622): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 9 issues
* ![MINOR][MINOR] 9 minor

Watch the comments in this conversation to review them.

#### 8 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [RenderManager.cpp#L504](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8b0ebc4309abcdc8c1e42664bbf260cc89d96d6a/src/Engine/Manager/RenderManager.cpp#L504): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L522](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8b0ebc4309abcdc8c1e42664bbf260cc89d96d6a/src/Engine/Manager/RenderManager.cpp#L522): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L541](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8b0ebc4309abcdc8c1e42664bbf260cc89d96d6a/src/Engine/Manager/RenderManager.cpp#L541): The function 'GetLenses' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L561](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8b0ebc4309abcdc8c1e42664bbf260cc89d96d6a/src/Engine/Manager/RenderManager.cpp#L561): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L579](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8b0ebc4309abcdc8c1e42664bbf260cc89d96d6a/src/Engine/Manager/RenderManager.cpp#L579): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L598](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8b0ebc4309abcdc8c1e42664bbf260cc89d96d6a/src/Engine/Manager/RenderManager.cpp#L598): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L620](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/8b0ebc4309abcdc8c1e42664bbf260cc89d96d6a/src/Engine/Manager/RenderManager.cpp#L620): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**